python/lsst/ts/ess/csc/controller_data_client.py

```diff
@@ -0,0 +1,449 @@
+                self.log.error(f"read_loop failed: {e}")
+                raise
+            except Exception:
+                self.log.exception("read_loop failed")
+                raise
+
+    async def read(self) -> dict:
+        """Read and unmarshal a json-encoded dict.
+
+        This may be a command acknowedgement or telemetry data.
+
+        Time out if reading takes longer than COMMUNICATE_TIMEOUT seconds.
+
+        Returns
+        -------
+        data : `dict`
+            The read data, after json-decoding it.
+        """
+        if not self.connected:
+            raise RuntimeError("Not connected.")
+        assert self.reader is not None  # make mypy happy
+
+        read_bytes = await asyncio.wait_for(
+            self.reader.readuntil(tcpip.TERMINATOR),
+            timeout=COMMUNICATE_TIMEOUT,
+        )
+        try:
+            data = json.loads(read_bytes.decode())
+        except json.decoder.JSONDecodeError as e:
+            raise RuntimeError(f"Could not parse {read_bytes!r} as json.") from e
+        if not isinstance(data, dict):
+            raise RuntimeError(
+                f"Could not parse {read_bytes!r} as a json-encoded dict."
+            )
+        return data
+
+    async def run_command(self, command: str, **parameters: Any) -> None:
+        """Write a command. Time out if it takes too long.
+
+        Parameters
+        ----------
+        command : `str`
+            The command to write.
+        **parameters : `dict`
+            Command parameters, as name=dict. For example::
+
+                configuration = {"devices": self.config.devices}
+
+        Raises
+        ------
+        ConnectionError
+            If not connected
+        asyncio.TimeoutError
+            If it takes more than COMMUNICATE_TIMEOUT seconds
+            to acquire the lock or write the data.
+        """
+        json_str = json.dumps({"command": command, "parameters": parameters})
+        await asyncio.wait_for(
+            self._basic_run_command(json_str), timeout=COMMUNICATE_TIMEOUT
+        )
+
+    async def _basic_run_command(self, json_str: str) -> None:
+        """Write a json-encoded command dict. Potentially wait forever.
+
+        Parameters
+        ----------
+        json_str : `str`
+            json-encoded dict to write. The dict should be of the form::
+
+                {"command": command_str, "parameters": params_dict}
+
+        Raises
+        ------
+        RuntimeError
+            If the command fails.
+        ConnectionError
+            If disconnected before command is acknowledged.
+        """
+        async with self.stream_lock:
+            if not self.connected:
+                raise ConnectionError("Not connected; cannot send the command.")
+            assert self.writer is not None  # make mypy happy
+
+            self.writer.write(json_str.encode() + tcpip.TERMINATOR)
+            await self.writer.drain()
+            while True:
```

Comment:
Will you be in the `while` loop forever?

Reply:
No. The task gets canceled when the application exits.